cluster/rpc_test.go

```diff
@@ -25,77 +34,158 @@ func (m *mockRegistry) Services(ctx context.Context) (map[string][]Node, error)
-					Port:    l.Addr().(*net.TCPAddr).Port,
-				},
-			},
-		},
+	go func() { _ = http.Serve(l, ts) }()
+
+	node := Node{
+		Address: "127.0.0.1",
+		Port:    l.Addr().(*net.TCPAddr).Port,
 	}
+	mock := newMockRegistry([]Node{node})
+	c, err := newClient("", "foo", &mock)
+	require.NoError(t, err)
+	defer c.Close()
 
-	client, err := NewClient("foo", &mock)
+	var reply string
+	err = c.Call("RPCTest.Call", "joe", &reply)
 	require.NoError(t, err)
-	require.NotNil(t, client)
+	require.Equal(t, "who's joe", reply)
 }
 
-func TestNewClient_with_no_available_server(t *testing.T) {
-	mock := mockRegistry{
-		services: map[string][]Node{},
-	}
+func TestNewConnectionBalancer_successul_inital_connect(t *testing.T) {
+	node, close := testRPCServer(t)
+	defer close()
 
-	client, err := NewClient("foo", &mock)
-	require.Error(t, err)
-	require.Nil(t, client)
+	mock := newMockRegistry([]Node{node})
+	c, err := newConnectionBalancer("", "foo", &mock)
+	require.NoError(t, err)
+	defer c.Close()
+
+	require.Equal(t, []Node{node}, c.getSelectedNodes())
+	require.Empty(t, c.errChan)
 }
 
-func TestNewClient_with_servers_failing_to_connect(t *testing.T) {
+func TestNewConnectionBalancer_with_no_available_server(t *testing.T) {
 	mock := mockRegistry{
-		services: map[string][]Node{
-			"foo": {
-				{Address: "127.0.0.1", Port: 1234},
-			},
-		},
+		nodeChan: make(chan []Node),
 	}
+	c, err := newConnectionBalancer("", "foo", &mock)
+	require.Error(t, err)
+	require.Nil(t, c)
+}
 
-	client, err := NewClient("foo", &mock)
+func TestNewConnectionBalancer_with_servers_failing_to_connect(t *testing.T) {
+	mock := newMockRegistry([]Node{{Address: "127.0.0.1", Port: 1234}})
+	c, err := newConnectionBalancer("", "foo", &mock)
 	require.Error(t, err)
-	require.Nil(t, client)
+	require.Nil(t, c)
 }
 
-func TestNodeToDial_uses_random_node(t *testing.T) {
-	mock := mockRegistry{
-		services: map[string][]Node{
-			"foo": {
-				{Address: "127.0.0.1", Port: 1234},
-				{Address: "127.0.0.1", Port: 3000},
-				{Address: "127.0.0.1", Port: 4321},
-			},
-		},
-	}
+func TestConnectionBalancer_watchForNewNodes(t *testing.T) {
+	node, close := testRPCServer(t)
+	defer close()
+	mock := newMockRegistry([]Node{node})
 
-	rand.Seed(1)
-	node, err := nodeToDial("foo", &mock)
+	c, err := newConnectionBalancer("", "foo", &mock)
 	require.NoError(t, err)
-	require.Equal(t, mock.services["foo"][2], node)
+	defer c.Close()
+
+	require.Equal(t, []Node{node}, c.getSelectedNodes())
+
+	go func() {
+		err := <-c.errChan
+		require.NoError(t, err)
+	}()
+
+	node2, close := testRPCServer(t)
+	defer close()
+	node3, close := testRPCServer(t)
+	defer close()
+	node4, close := testRPCServer(t)
+	defer close()
+
+	t.Run("test when more than max connections of nodes get sent", func(t *testing.T) {
+		go func() {
+			mock.nodeChan <- []Node{node, node2, node3, node4}
+		}()
+		time.Sleep(4 * time.Millisecond)
+		require.Equal(t, []Node{node4, node, node2}, c.getSelectedNodes())
```

Comment:
anyone know a better way than a sleep to wait for the `watchForNewNodes` to update the node list?

Reply:
I like this idea, like a updatedNodes channel